src/Sarif.Viewer.VisualStudio/ErrorList/ColumnFilterer.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.TableControl;
+
+namespace Microsoft.Sarif.Viewer.ErrorList
+{
+    // This class allows the caller to filter specified values from specified columns.
+    // It keeps track of which column/value pairs have been filtered, so that it only
+    // does so once per column/value pair. The idea is that if the user ever checks the
+    // box in the filter UI to show a certain value in a certain column, we should never
+    // hide that column again.
+    internal class ColumnFilterer
+    {
+        // The set of column/value pairs that have been filtered so far.
+        private readonly HashSet<FilteredColumnValue> filteredColumnValues = new HashSet<FilteredColumnValue>();
+
+        private IWpfTableControl errorListTableControl;
+
+        private IWpfTableControl ErrorListTableControl
+        {
+            get
+            {
+                if (this.errorListTableControl == null)
+                {
+                    var componentModel = (IComponentModel)Package.GetGlobalService(typeof(SComponentModel));
```

Comment:
My expectation is you can also do this via:
    var errorList = Package.GetGlobalService(typeof(SVsErrorList)) as IErrorList;
    return errorList?.TableControl;

Reply:
NM, I see the implementation of IErrorListTableControlProvider.

I'm not sure whether there's any value to the extra level of indirection.

Reply:
I originally thought about injecting the `IWpfTableControl` into this class, so I wanted a provider for it, but yes, it's overkill.

---
In reply to: [537999712](https://github.com/microsoft/sarif-visualstudio-extension/pull/313#discussion_r537999712) [](ancestors = 537999712)